src/helper/helper.js

```diff
@@ -150,6 +161,59 @@ class DataLayerHelper {
     merge_(this.model_, /** @type {!Object<*>} */ (this.dataLayer_[0]));
   }
 
+
+  /**
+   * Register a function to respond to events with a certain name called by
+   * the command API by storing it in a map. The function will be called an
+   * time the commandAPI is called with first parameter the string name.
+   *
+   * @param {string} name The string which should be passed into the command API
+   *     to call the processor.
+   * @param {function(*):!Object} processor The function to register to be
+   *     called later.
+   *     So long as this function is not an arrow function, it can access the
+   *     abstract model interface by using the this keyword. It is recommended
+   *     not to modify the state within the function using this.set.
+   *     Changes to the model should only be achieved by the return value, a
+   *     dict whose values will automatically be merged into the model.
+   * @this {DataLayerHelper}
+   */
+  registerProcessor(name, processor) {
+    if (!(name in this.commandProcessors_)) {
+      this.commandProcessors_[name] = [];
+    }
+    this.commandProcessors_[name].push(processor);
+  }
+
+  /**
+   * Applies the given command to the value in the dataLayer with the given key.
+   * If a processor for the command has been registered, the processor function
+   * will be invoked with any arguments passed in.
+   *
+   * @param {!Array<*>} args The arguments object containing the command
+   *     to execute and optional arguments for the processor.
+   * @return {!Array<!Object<*>>} states The updates requested to
+   *     the model state, in the order they should be processed.
+   * @private
+   */
+  processArguments_(args) {
+    // Run all registered processors associated with this command.
+    const states = [];
+    const name = args[0];
+    if (this.commandProcessors_[name]) {
+      // Cache length so as not to run processors registered
```

Comment:
I'm not sure I understand this comment? Are you concerned that someone else might add to "commandProcessors_" during the loop?

Reply:
```
   it('calls registered functions only when the commandAPI ' +
        'fires the appropriate event', () => {
      dataLayerHelper.registerProcessor('process', () => {
        spy('a');
        dataLayerHelper.registerProcessor('process', () => {
          spy('b');
        });
      });
      commandAPI('process');

      expect(spy.calls.allArgs()).toEqual([['a']]);
    });
```
Yes, without this the above test would end with [['a'], ['b']]. I think that it is easier to reason about the claim 'commandAPI('process') calls all processors registered at the time' than the claim that it 'calls all processors registered before it is finished processing'. Though I don't think there are many (any?) use cases for this anyways. I added a bit to the comment there for clarity.